scipy/optimize/tests/test_linprog.py

```diff
@@ -1675,6 +1676,68 @@ def test_crossover(self):
         # there should be nonzero crossover iterations for IPM (only)
         assert_equal(res.crossover_nit == 0, self.method != "highs-ipm")
 
+    def test_marginals(self):
+        # Ensure lagrange multipliers are correct by comparing the derivative
+        # w.r.t. b_ub/b_eq/ub/lb to the reported duals.
+        c, A_ub, b_ub, A_eq, b_eq, bounds = very_random_gen(seed=0)
+        res = linprog(c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq,
+                      bounds=bounds, method=self.method, options=self.options)
+        lb, ub = bounds.T
+
+        # sensitivity w.r.t. b_ub
+        def f_bub(x):
+            return linprog(c, A_ub, x, A_eq, b_eq, bounds,
+                           method=self.method).fun
+
+        dfdbub = approx_derivative(f_bub, b_ub, method='3-point', f0=res.fun)
+        assert_allclose(res.marginals.ineqlin, dfdbub)
+
+        # sensitivity w.r.t. b_eq
+        def f_beq(x):
+            return linprog(c, A_ub, b_ub, A_eq, x, bounds,
+                           method=self.method).fun
+
+        dfdbeq = approx_derivative(f_beq, b_eq, method='3-point', f0=res.fun)
+        assert_allclose(res.marginals.eqlin, dfdbeq)
+
+        # sensitivity w.r.t. lb
+        def f_lb(x):
+            bounds = np.array([x, ub]).T
+            return linprog(c, A_ub, b_ub, A_eq, b_eq, bounds,
+                           method=self.method).fun
+
+        with np.errstate(invalid='ignore'):
+            # approx_derivative has trouble where lb is infinite
+            dfdlb = approx_derivative(f_lb, lb, method='3-point', f0=res.fun)
+            dfdlb[~np.isfinite(lb)] = 0
+
+        assert_allclose(res.marginals.lower, dfdlb)
+
+        # sensitivity w.r.t. ub
+        def f_ub(x):
+            bounds = np.array([lb, x]).T
+            return linprog(c, A_ub, b_ub, A_eq, b_eq, bounds,
+                           method=self.method).fun
+
+        with np.errstate(invalid='ignore'):
+            dfdub = approx_derivative(f_ub, ub, method='3-point', f0=res.fun)
+            dfdub[~np.isfinite(ub)] = 0
+
+        assert_allclose(res.marginals.upper, dfdub)
+
+    def test_complementary_slackness(self):
+        # Ensure that the complementary slackness condition is satisfied.
+        c, A_ub, b_ub, A_eq, b_eq, bounds = very_random_gen(seed=42)
+        res = linprog(c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq,
+                      bounds=bounds, method=self.method, options=self.options)
+
+        resid = (-c + A_ub.T @ res.marginals.ineqlin +
+                 A_eq.T @ res.marginals.eqlin +
+                 res.marginals.upper +
+                 res.marginals.lower)
```

Comment:
Could you add a reference for this equation as a comment?

Reply:
Split up into two KKT tests: dual feasibility (what this test really was) and complementary slackness.  Added references to both.

Reply:
I realize now that this was just the identity I saw in Matlab's documentation:
https://www.mathworks.com/help/optim/ug/linprog.html#buusznx-lambda

Why are the signs different from that, though?

Reply:
I believe it has to do with the sign convention of the Lagrangian.  As noted in the note below the equation on that MATLAB doc page:
> This sign convention matches that of nonlinear solvers (see Constrained Optimality Theory). However, this sign is the opposite of the sign in much linear programming literature, so a linprog Lagrange multiplier is the negative of the associated "shadow price." 

Since we're using the shadow price to calculate the dual feasibility, we write: 

    c.T @ x + ineqlin.T @ (b_ub - A_ub @ x) + eqlin.T @ (b_eq - A_eq @ x) + upper.T @ (ub - x) + lower.T @ (lb - x)

This reverses the sign where the Lagrangian multipliers show up (i.e. our `b_ub - A_ub @ x` vs MATLAB's `A_ub @ x - b_ub`).  Taking the derivative w.r.t. x of the above gives the signs used in the test.  So I'm pretty sure this this is equivalent, we're just "shadow prices" where they're using Lagrange multipliers with an assumed sign convention.

Reply:
I've added a note about sign convention to the `Notes` section of the docstrings